heartbeat/skima/core.go

```diff
@@ -0,0 +1,168 @@
+	return func(actual common.MapStr) SchemaValidationResults {
+		results := make([]SchemaValidationResults, len(validators))
+		for idx, validator := range validators {
+			results[idx] = validator(actual)
+		}
+
+		combined := SchemaValidationResults{}
+		for _, r := range results {
+			r.eachResult(func(path string, vr ValueResult) {
+				combined.recordResult(path, vr)
+			})
+		}
+		return combined
+	}
+}
+
+// Strict is used when you want any unspecified keys that are encountered to be considered errors.
+func Strict(laxValidator SchemaValidator) SchemaValidator {
+	return func(actual common.MapStr) SchemaValidationResults {
+		validatedResults := laxValidator(actual)
+
+		// The inner workings of this are a little weird
+		// We use a hash of dotted paths to track the results
+		// We can check if a key had a test associated with it by looking up the laxValidator
+		// result data
+		// What's trickier is intermediate maps, maps don't usually have explicit tests, they usually just have
+		// their properties tested.
+		// This method counts an intermediate map as tested if a subkey is tested.
+		// Since the datastructure we have to search is a flattened hashmap of the original map we take that hashmap
+		// and turn it into a sorted string array, then do a binary prefix search to determine if a subkey was tested.
+		// It's a little weird, but is fairly efficient. We could stop using the flattened map as a datastructure, but
+		// that would add complexity elsewhere. Probably a good refactor at some point, but not worth it now.
+		validatedPaths := sort.StringSlice{}
+		for k, _ := range validatedResults {
+			validatedPaths = append(validatedPaths, k)
+		}
+		validatedPaths.Sort()
+
+		walk(actual, func(
+			key string,
+			value interface{},
+			currentMap common.MapStr,
+			rootMap common.MapStr,
+			path []string,
+			dottedPath string) {
+
+			_, validatedExactly := validatedResults[dottedPath]
+			if validatedExactly {
+				return // This key was tested, passes strict test
+			}
+
+			// Search returns the point just before an actual match (since we ruled out an exact match with the cheaper
+			// hash check above. We have to validate the actual match with a prefix check as well
+			matchIdx := validatedPaths.Search(dottedPath)
+			if matchIdx < len(validatedPaths) && strings.HasPrefix(validatedPaths[matchIdx], dottedPath) {
+				return
+			}
+
+			validatedResults.recordResult(dottedPath, StrictFailureVR)
+		})
+
+		return validatedResults
+	}
+}
+
+// Test takes the output from a SchemaValidator invocation and runs test assertions on the result.
+// If you are using this library for testing you will probably want to run Test(t, Schema(Map{...})(actual)) as a pattern.
+func Test(t *testing.T, r SchemaValidationResults) SchemaValidationResults {
+	r.Errors().eachResult(func(p string, vr ValueResult) {
+		msg := fmt.Sprintf("@path '%s': %s", p, vr.message)
+		assert.True(t, vr.valid, msg)
+	})
+	return r
+}
```

Comment:
How about changing this to `DetailedErrors` which returns an error wrapping multiple errors.

```
type multiErr []error

type (m *multiErr) Error() string {

Reply:
I like the idea, I'd still like to keep around the `map[string][]ValueResult` structure for future use, but for testing it is overkill.